File: java/src/main/java/io/cucumber/gherkin/Token.java
package io.cucumber.gherkin;

import io.cucumber.messages.types.Location;
import io.cucumber.messages.types.StepKeywordType;
import org.jspecify.annotations.Nullable;

import java.util.List;

import static java.util.Objects.requireNonNull;

final class Token {
    final @Nullable Line line;
    final boolean eof;
    Parser.@Nullable TokenType matchedType;
    @Nullable String matchedKeyword;
    @Nullable String matchedText;
    @Nullable List<LineSpan> matchedItems;
    int matchedIndent;
    @Nullable String matchedLanguage;
    @Nullable StepKeywordType keywordType;
    Location location;

    private Token(@Nullable Line line, Location location) {
        this.line = line;
        this.location = location;
        this.eof = line == null;
    }

    static Token createEOF(Location location) {
        requireNonNull(location);
        return new Token(null, location);
    }

    static Token createGherkinLine(String text, Location location) {
        requireNonNull(text);
        requireNonNull(location);
        return new Token(new Line(text), location);
    }

    boolean isEOF() {
        return eof;
    }

    String getTokenValue() {
        return line == null ? "EOF" : line.getText();
    }

    @Override
    public String toString() {
        return "%s: %s/%s".formatted(matchedType, matchedKeyword, matchedText);
    }

    String getRequiredMatchedKeyword() {
        return requireNonNull(matchedKeyword);
    }

    String getRequiredMatchedText() {
        return requireNonNull(matchedText);
    }

    String getRequiredMatchedLanguage() {
        return requireNonNull(matchedLanguage);
    }

    Parser.TokenType getRequiredMatchedType() {
        return requireNonNull(matchedType);
    }

    Line getRequiredLine() {
        return requireNonNull(line);
    }
}

File: dart/lib/src/parser/token.dart
import 'package:cucumber_gherkin/src/parser/gherkin_line.dart';
import 'package:cucumber_gherkin/src/parser/parser.g.dart';
import 'package:cucumber_messages/cucumber_messages.dart' as messages;

class Token {
  Token(this.line, this.location);

  Token.eof(this.location) : line = null;

  final GherkinLine? line;

  messages.Location location;

  TokenType matchedType = TokenType.none;

  String matchedKeyword = '';

  messages.StepKeywordType? matchedKeywordType;

  String matchedText = '';

  Iterable<GherkinLineSpan> matchedItems = const <GherkinLineSpan>[];

  String? matchedLanguage;

  bool get isEof => line == null;

  String get tokenValue => isEof ? 'EOF' : line!.getLineText(-1);
}
